x/valset-pref/validator_set.go

```diff
@@ -170,49 +214,55 @@ func (k Keeper) PreformRedelegation(ctx sdk.Context, delegator sdk.AccAddress, e
 	}
 
 	// calculate the difference between two sets
-	var diffValSet []*valSet
+	var diffValSets []*valSet
+	// Matt: Does this mean that the diffAmount can be negative?
 	for i, newVals := range existingValSet {
 		diffAmount := newVals.amount.Sub(newValSet[i].amount)
 
 		diff_val := valSet{
 			valAddr: newVals.valAddr,
 			amount:  diffAmount,
 		}
-		diffValSet = append(diffValSet, &diff_val)
+		diffValSets = append(diffValSets, &diff_val)
 	}
 
 	// Algorithm starts here
-	for _, diff_val := range diffValSet {
-		for diff_val.amount.TruncateDec().GT(sdk.NewDec(0)) {
-			source_val := diff_val.valAddr
-			target_val, idx := k.FindMin(diffValSet, source_val)
+	// Matt: we need more detailed documnetaiton on the algorithm here
```

Comment:
the documentation got a bit too large so i added it in `## Redelegate algorithm logic pseudocode` I can link it here